### ISSUE TITLE
Alerting: Optimistic locking in the Alertmanager config table causes write lock contention in sqlite

### ISSUE THREAD

**Issue by alexweav** [grafana/grafana#51356]
**What happened**:

### Background

The Alertmanager config table is only ever inserted to. Making any change results in a new row being inserted, with an incremented ID. To get the current config for an org, we currently query `WHERE id=(SELECT MAX(id) from alert_configuration WHERE org_id = ?)`

According to git and PR history, this was originally done for recordkeeping purposes. However, we never actually use the old configs at all currently.

To perform optimistic locking on this table, we have to take a convoluted approach since this table is INSERT-only. Right now, we `INSERT WHERE EXISTS` a record such that `id = (SELECT MAX(id) ...` and `configuration_hash = <last known hash>`. This makes sure that we only insert if the highest known record has the hash we expect.

We can't do this using a separate query and write in the same transaction, either. This doesn't produce the expected results in all the databases we support. For example, most Postgres installations only have support Read Committed transaction isolation, which means that concurrent transactions can bypass the optimistic locking entirely, if it's implemented with two commands. One command is the only way to make sure that we keep our safety guarantees against all databases.

### The Problem

The above query acquires a full table lock on `alert_configuration`. Since many alerting resources are all stored in alert configs, this table gets updated heavily. Many requests in a row are subject to write lock contention, and take progressively longer.

**What you expected to happen**:

In my testing, I used the terraform provider's test suite as a substitute for a "big burst of traffic." It sends lots of write requests over and over, for every resource.

Some request latencies I saw as the test suite ran:
```
93ms
151ms
321ms
488ms
971ms
1872ms
4543ms
7981ms
14922ms
34870ms
117724ms
```

Right now we tell customers to not use sqlite for heavier production workloads. This is true to a point, but the above is well beyond that threshold IMO. Reducing the lock acquisition here makes the experience much better for initial users. This should also ideally improve perf in other databases as well.

**How to reproduce it (as minimally and precisely as possible)**:

Run the terraform provider tests, as I did above. It's a little intermittent, since locking problems are never that easy to reproduce reliably.

**Anything else we need to know?**:

**Environment**:
- Grafana version: main
- Data source type & version: n/a
- OS Grafana is installed on: Linux
- User OS & Browser: Firefox
- Grafana plugins: n/a
- Others: Sqlite database.

**Comment by alexweav** [comment]
In lieu of a design doc, I see three ways of approaching this problem:

1. Do nothing.

Continue to claim that serious alerting users should move off sqlite. Masks even serious perf problems, and is a very bad initial user experience in some cases.

2. UPDATE rather than INSERT

I tried replacing the above query with an in-place UPDATE:
```
<plain old update> WHERE org_id = ? AND configuration_hash = ? AND id = (SELECT MAX(id) FROM alert_configuration WHERE org_id = ?)
```

This solved the locking problem and requests were all sub-100ms under even quite a lot of load.

The downside is that this no longer tracks the history all AM config versions in the database. We stop grabbing history that we currently grab.

3. UPDATE, with audit logging

This is an extension of option 2 where we implement an audit-logging system for AM configs. The "active" config for each org is stored in a dedicated row which gets updated. Every time we write to it, we push the old version (or a summary of the changes) to another table, which is a simple insert. This allows us to keep our history tracking, but also have fast updates.

**Comment by alexweav** [comment]
IMHO, my vote is currently for option 2. We don't do anything with the history today. There is no code that accesses these old configs, and we've never manually dug them out in an escalation. The audit logging stuff is YAGNI - let's implement it when we actually need it.

**Comment by JacobsonMT** [comment]
I know this is a problem we should fix asap but I'm a bit hesitant to remove the `alert_configuration` history without some sort of replacement/temp standin. I've seen it mentioned in a few issues so far where it was used to rollback after something clobbered a user's current config. For example, the sqlite db lock issue I recently worked on.

The downside of an overwritten alermanager config can be tremendous considering how much info is stored in there.

**Comment by JohnnyQQQQ** [comment]
Fixed by #51461
